src/v/cluster/rm_stm.cc

```diff
@@ -250,7 +250,7 @@ ss::future<checked<model::term_id, tx_errc>> rm_stm::do_begin_tx(
     auto fence_it = _log_state.fence_pid_epoch.find(pid.get_id());
     auto is_new_pid = fence_it == _log_state.fence_pid_epoch.end();
     if (is_new_pid || pid.get_epoch() > fence_it->second) {
-        if (!is_new_pid && pid.get_epoch() > fence_it->second) {
```

Comment:
just making sure this isn't a mistake because the commit message only alludes to adding constraints:

> adding an extra check to protect from incorrect clients.

Reply:
it isn't a mistake because the are equivalent: 

```
if (a || b) {
  if (!a && b) {
  }
}
```

is equivalent to

```
if (a || b) {
  if ((!a && b) && (a || b)) {
  }
}
```

since `(!a && b) && (a || b) == ((!a && b) && a) || ((!a && b) && b) == (!a && b)` we may rewrite it further

```
if (a || b) {
  if (!a && b) {
  }
}
```

let's repeat the same process with

```
if (a || b) {
  if (!a) {
  }
}
```

`((a || b) && !a) == (a&&!a) || (b && !a) == b && a!`

so

```
if (a || b) {
  if (!a && b) {
  }
}
```

is the same as 

```
if (a || b) {
  if (!a) {
  }
}
```